src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -367,6 +372,23 @@ private static string GetPreferredUsernameFromIdToken(bool isAdfsAuthority, IdTo
 
             CacheInfoTelemetry cacheInfoTelemetry = CacheInfoTelemetry.None;
 
+            // Verify appropriate scenario for OBO tokens in long running process
+            if (requestParams.ApiId == ApiEvent.ApiIds.AcquireTokenOnBehalfOf)
+            {
+                // Validates correct AcquireTokenInLongRunningProcess call (user assertion is null in this case)
+                if (requestParams.UserAssertion == null && !string.IsNullOrEmpty(requestParams.OboCacheKey) && tokenCacheItems.Count == 0)
+                {
+                    logger.Error("[FindAccessTokenAsync] AcquireTokenInLongRunningProcess was called and OBO token was not found in the cache.");
+                    throw new MsalClientException(MsalError.OboCacheKeyNotInCacheError, MsalErrorMessage.OboCacheKeyNotInCache);
+                }
+                // Validates correct InitiateLongRunningProcessInWebAPI call (user assertion is specified in this case)
+                if (requestParams.UserAssertion != null && !string.IsNullOrEmpty(requestParams.OboCacheKey) && tokenCacheItems.Count > 0)
```

Comment:
I think it will be simpler to override the existing token instead of throwing? If the key is the user id and the user re-logs in. CC @jmprieur for exception behavior.